scripts/generateReadme.js

```diff
@@ -0,0 +1,88 @@
+import * as child_process from "child_process";
+
+const TEMPLATE_FILE = "README.template.md";
+const OUTPUT_FILE = "README.md";
+
+/**
+ * Determine version from package.json
+ */
+const packageJson = JSON.parse(await fs.readFile("package.json", "utf8"));
+const version = packageJson.version;
+
+/**
+ * Run test and get coverage
+ *
+ * If the command exits with 0, the tests pass. Otherwise, the tests have failed.
+ * The command output (regardless of exit code) will return the coverage report, which looks like this:
+ *
+ * --------------|---------|----------|---------|---------|-------------------
+ * File          | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
+ * --------------|---------|----------|---------|---------|-------------------
+ * All files     |     100 |      100 |     100 |     100 |
+ *  RateLimit.js |     100 |      100 |     100 |     100 |
+ * --------------|---------|----------|---------|---------|-------------------
+ *
+ * The coverage percentage is the average of all columns for "All files".
+ */
+const testsResult = await new Promise((resolve, reject) => {
+    const child = child_process.spawn("npm", ["test"]);
+    const data = [];
+    child.stdout.on("data", (chunk) => data.push(chunk));
+    child.on("exit", (code) => {
+        // Get coverage percentage
+        const output = data.join("");
+        const coverage = (output.match(/----(?:\n|.)+/g) ?? []).pop() ?? "";
+        const percentages = coverage.match(/All files\s*?\|\s*?(\d+)\s*?\|\s*?(\d+)\s*?\|\s*?(\d+)\s*?\|\s*?(\d+)\s*?/) ?? [];
+        const averageCoverage = percentages.slice(1).reduce((a, b) => a + parseInt(b, 10), 0) / 4;
+        resolve({coverage: averageCoverage, testsPass: code === 0});
+    });
+});
+
+/**
+ * Coverage colors
+ */
+const coverageColors = ["16a34a", "84cc16", "eab308", "f59e0b", "f97316", "ef4444"];
+
+/**
+ * Determine if the build passes by the exit code of `npm run build`
+ */
+const buildPass = await new Promise((resolve, reject) => {
+    const child = child_process.spawn("npm", ["run", "build"]);
+    child.on("exit", (code) => resolve(code === 0));
+});
+
+/**
+ * Variables to replace in the template file.
+ *
+ * {{variable_name}}
+ */
+const variables = {
+    "shield:version": `![version: ${version}](https://img.shields.io/badge/version-${version}-%233b82f6)`,
+    "shield:tests": `![test: ${testsResult.testsPass ? "passing" : "failing"}](https://img.shields.io/badge/tests-${testsResult.testsPass ? "passing" : "failing"}-${testsResult.testsPass ? "%2316a34a" : "%23ef4444"})`,
+    "shield:coverage": `![coverage: ${testsResult.coverage}%](https://img.shields.io/badge/coverage-${testsResult.coverage}%25-%23${coverageColors.reverse()[Math.floor(testsResult.coverage * (coverageColors.length - 1) / 100)]})`,
+    "shield:build": `![build: ${buildPass ? "passing" : "failing"}](https://img.shields.io/badge/build-${buildPass ? "passing" : "failing"}-${buildPass ? "%2316a34a" : "%23ef4444"})`,
```

Comment:
I don't understand the usage of `coverageColors` constant on this case since you are manually setting a value if true or false maybe you could get from array.

Reply:
Coverage is a percentage from 0 to 100. I set a colour from red to green depending on the percentage. `coverageColors` are the colours in linear scale and then I just assign each colour to a specific margin

Reply:
I did get that but you are manually setting a color here look:
``    "shield:build": `![build: ${buildPass ? "passing" : "failing"}](https://img.shields.io/badge/build-${buildPass ? "passing" : "failing"}-${buildPass ? "%2316a34a" : "%23ef4444"})`,``
instead of `%2316a34a` you could `` `${%23coverageColors[0]}` ``
instead of `%23ef4444` you could `` `${%23coverageColors[5]}` ``
and so on

Reply:
Ah, I see. The thing is, this is a different colour. For `tests: passing/failing` and `build: passing/failing` I only use red and green (just simply for true/false). Meanwhile, the `coverageColours` is a *colour scale*, e.g. 100% green, 70% orange, 30% red, 5% dark red, etc. The colour scale can actually have any number of elements (what colour to use is calculated automatically based on the length) and overall it just illustrates something different.

Reply:
Btw this should have been discussed on #2